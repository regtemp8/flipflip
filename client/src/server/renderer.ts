import createCache, { type EmotionCache } from '@emotion/cache'
import { configureStore } from '@reduxjs/toolkit'
import { AppStorage, SystemConstants } from 'flipflip-common'
import createAppReducer from '../store/app/slice'
import createAudioReducer from '../store/audio/slice'
import createCaptionScriptReducer from '../store/captionScript/slice'
import createClipReducer from '../store/clip/slice'
import createLibrarySourceReducer from '../store/librarySource/slice'
import createPlaylistReducer from '../store/playlist/slice'
import createSceneReducer from '../store/scene/slice'
import createSceneGroupReducer from '../store/sceneGroup/slice'
import createTagReducer from '../store/tag/slice'
import playersSlice from '../store/player/slice'
import videoClipperSlice from '../store/videoClipper/slice'
import scenePickerSlice from '../store/scenePicker/slice'
import captionScriptorSlice from '../store/captionScriptor/slice'
import sceneDetailSlice from '../store/sceneDetail/slice'
import audioLibrarySlice from '../store/audioLibrary/slice'
import sourceScraperSlice from '../store/sourceScraper/slice'
import createConstantsReducer from '../store/constants/slice'
import { fromAppStorage } from '../store/app/convert'
import createDisplayReducer from '../store/display/slice'
import createDisplayViewReducer from '../store/displayView/slice'
import createScenePlaylistItemReducer from '../store/scenePlaylistItem/slice'
import createDisplayPlaylistItemReducer from '../store/displayPlaylistItem/slice'

export function createEmotionCache(): EmotionCache {
  return createCache({ key: 'css' })
}

export function createReduxStore(
  appStorage?: AppStorage,
  constants?: SystemConstants
) {
  const state = appStorage != null ? fromAppStorage(appStorage) : undefined
  return configureStore({
    reducer: {
      app: createAppReducer(state?.app),
      audio: createAudioReducer(state?.audio),
      captionScript: createCaptionScriptReducer(state?.captionScript),
      clip: createClipReducer(state?.clip),
      librarySource: createLibrarySourceReducer(state?.librarySource),
      playlist: createPlaylistReducer(state?.playlist),
      displayPlaylistItem: createDisplayPlaylistItemReducer(
        state?.displayPlaylistItem
      ),
      scenePlaylistItem: createScenePlaylistItemReducer(
        state?.scenePlaylistItem
      ),
      scene: createSceneReducer(state?.scene),
      sceneGroup: createSceneGroupReducer(state?.sceneGroup),
      display: createDisplayReducer(state?.display),
      displayView: createDisplayViewReducer(state?.displayView),
      tag: createTagReducer(state?.tag),
      constants: createConstantsReducer(constants),
      // components
      players: playersSlice,
      videoClipper: videoClipperSlice,
      scenePicker: scenePickerSlice,
      captionScriptor: captionScriptorSlice,
      sceneDetail: sceneDetailSlice,
      audioLibrary: audioLibrarySlice,
      sourceScraper: sourceScraperSlice
    }
  })
}
