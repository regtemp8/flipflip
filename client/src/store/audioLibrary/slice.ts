import { createSlice, type PayloadAction } from '@reduxjs/toolkit'

interface AudioLibraryState {
  error: boolean
  loadingMetadata: boolean
  loadingSources: boolean
  drawerOpen: boolean
  selectedTags: string[]
  openMenu?: string
  playlistID?: number
  importURL?: string
}

export const initialAudioLibraryState: AudioLibraryState = {
  error: false,
  loadingMetadata: false,
  loadingSources: false,
  drawerOpen: false,
  selectedTags: []
}

export const audioLibrarySlice = createSlice({
  name: 'audioLibrary',
  // `createSlice` will infer the state type from the `initialState` argument
  initialState: initialAudioLibraryState,
  reducers: {
    setError: (state, action: PayloadAction<boolean>) => {
      state.error = action.payload
    },
    setLoadingMetadata: (state, action: PayloadAction<boolean>) => {
      state.loadingMetadata = action.payload
    },
    setLoadingSources: (state, action: PayloadAction<boolean>) => {
      state.loadingSources = action.payload
    },
    setOpenMenu: (state, action: PayloadAction<string | undefined>) => {
      state.openMenu = action.payload
    },
    setPlaylistID: (state, action: PayloadAction<number | undefined>) => {
      state.playlistID = action.payload
    },
    setDrawerOpen: (state, action: PayloadAction<boolean>) => {
      state.drawerOpen = action.payload
    },
    setImportURL: (state, action: PayloadAction<string | undefined>) => {
      state.importURL = action.payload
    },
    setSelectedTags: (state, action: PayloadAction<string[]>) => {
      state.selectedTags = action.payload
    },
    setAudioLibraryPlaylistID: (state, action: PayloadAction<string>) => {
      state.playlistID = Number(action.payload)
    }
  }
})

export const {
  setError,
  setLoadingMetadata,
  setLoadingSources,
  setOpenMenu,
  setPlaylistID,
  setDrawerOpen,
  setImportURL,
  setSelectedTags,
  setAudioLibraryPlaylistID
} = audioLibrarySlice.actions

export default audioLibrarySlice.reducer
