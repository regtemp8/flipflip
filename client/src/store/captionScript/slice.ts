import { createSlice, type PayloadAction } from '@reduxjs/toolkit'
import type CaptionScript from './CaptionScript'
import {
  type EntryState,
  type EntryUpdate,
  setEntrySlice,
  setEntry,
  updateEntry
} from '../EntryState'
import { type FontSettingsType } from './FontSettings'
import { getCaptionScript, getCaptionScriptFontSettings } from './selectors'

export interface FontSettingsUpdate<T> extends EntryUpdate<T> {
  type: FontSettingsType
}

export const initialCaptionScriptState: EntryState<CaptionScript> = {
  name: 'captionScriptSlice',
  nextID: 1,
  entries: {}
}
export const captionScriptSlice = createSlice({
  name: 'captionScripts',
  // `createSlice` will infer the state type from the `initialState` argument
  initialState: initialCaptionScriptState,
  reducers: {
    setCaptionScriptSlice: (
      state,
      action: PayloadAction<EntryState<CaptionScript>>
    ) => {
      setEntrySlice(state, action.payload)
    },
    setCaptionScript: (state, action: PayloadAction<CaptionScript>) => {
      setEntry(state, action.payload)
    },
    updateCaptionScript: (state, action: PayloadAction<CaptionScript>) => {
      updateEntry(state, action.payload)
    },
    setCaptionScriptScript: (
      state,
      action: PayloadAction<EntryUpdate<string>>
    ) => {
      getCaptionScript(state, action.payload.id).script = action.payload.value
    },
    setCaptionScriptURL: (
      state,
      action: PayloadAction<EntryUpdate<string>>
    ) => {
      getCaptionScript(state, action.payload.id).url = action.payload.value
    },
    setCaptionScriptStopAtEnd: (
      state,
      action: PayloadAction<EntryUpdate<boolean>>
    ) => {
      getCaptionScript(state, action.payload.id).stopAtEnd =
        action.payload.value
    },
    setCaptionScriptNextSceneAtEnd: (
      state,
      action: PayloadAction<EntryUpdate<boolean>>
    ) => {
      getCaptionScript(state, action.payload.id).nextSceneAtEnd =
        action.payload.value
    },
    setCaptionScriptSyncWithAudio: (
      state,
      action: PayloadAction<EntryUpdate<boolean>>
    ) => {
      getCaptionScript(state, action.payload.id).syncWithAudio =
        action.payload.value
    },
    setCaptionScriptOpacity: (
      state,
      action: PayloadAction<EntryUpdate<number>>
    ) => {
      getCaptionScript(state, action.payload.id).opacity = action.payload.value
    },
    setCaptionScriptFontSettingsBorder: (
      state,
      action: PayloadAction<FontSettingsUpdate<boolean>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).border = action.payload.value
    },
    setCaptionScriptFontSettingsColor: (
      state,
      action: PayloadAction<FontSettingsUpdate<string>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).color = action.payload.value
    },
    setCaptionScriptFontSettingsBorderColor: (
      state,
      action: PayloadAction<FontSettingsUpdate<string>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).borderColor = action.payload.value
    },
    setCaptionScriptFontSettingsFontFamily: (
      state,
      action: PayloadAction<FontSettingsUpdate<string>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).fontFamily = action.payload.value
    },
    setCaptionScriptFontSettingsFontSize: (
      state,
      action: PayloadAction<FontSettingsUpdate<number>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).fontSize = action.payload.value
    },
    setCaptionScriptFontSettingsBorderPx: (
      state,
      action: PayloadAction<FontSettingsUpdate<number>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).borderpx = action.payload.value
    },
    setCaptionScriptTags: (
      state,
      action: PayloadAction<EntryUpdate<number[]>>
    ) => {
      state.entries[action.payload.id].tags = action.payload.value
    },
    setCaptionScriptToggleTag: (
      state,
      action: PayloadAction<EntryUpdate<number>>
    ) => {
      const script = state.entries[action.payload.id]
      const index = script.tags.indexOf(action.payload.value)
      if (index === -1) {
        script.tags.push(action.payload.value)
      } else {
        script.tags.splice(index, 1)
      }
    },
    setCaptionScriptMarked: (
      state,
      action: PayloadAction<EntryUpdate<boolean>>
    ) => {
      getCaptionScript(state, action.payload.id).marked = action.payload.value
    }
  }
})

export const {
  setCaptionScriptSlice,
  setCaptionScript,
  updateCaptionScript,
  setCaptionScriptScript,
  setCaptionScriptURL,
  setCaptionScriptStopAtEnd,
  setCaptionScriptNextSceneAtEnd,
  setCaptionScriptSyncWithAudio,
  setCaptionScriptOpacity,
  setCaptionScriptFontSettingsBorder,
  setCaptionScriptFontSettingsColor,
  setCaptionScriptFontSettingsBorderColor,
  setCaptionScriptFontSettingsFontFamily,
  setCaptionScriptFontSettingsFontSize,
  setCaptionScriptFontSettingsBorderPx,
  setCaptionScriptTags,
  setCaptionScriptToggleTag,
  setCaptionScriptMarked
} = captionScriptSlice.actions

export default captionScriptSlice.reducer
