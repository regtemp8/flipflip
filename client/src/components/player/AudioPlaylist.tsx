import React, { MouseEvent, useEffect, useState } from 'react'
import Sortable from 'react-sortablejs'

import {
  Avatar,
  Badge,
  Chip,
  Fab,
  IconButton,
  List,
  ListItem,
  ListItemAvatar,
  ListItemIcon,
  ListItemSecondaryAction,
  ListItemText,
  type Theme,
  Tooltip
} from '@mui/material'

import { makeStyles } from 'tss-react/mui'

import AddIcon from '@mui/icons-material/Add'
import AudiotrackIcon from '@mui/icons-material/Audiotrack'
import BuildIcon from '@mui/icons-material/Build'
import ClearIcon from '@mui/icons-material/Clear'
import DeleteIcon from '@mui/icons-material/Delete'
import RepeatIcon from '@mui/icons-material/Repeat'
import RepeatOneIcon from '@mui/icons-material/RepeatOne'
import ShuffleIcon from '@mui/icons-material/Shuffle'

import { arrayMove, getTimestamp, randomizeList } from '../../data/utils'
import { RP } from 'flipflip-common'
import AudioControl from './AudioControl'
import SourceIcon from '../library/SourceIcon'
import { setAudioVolume } from '../../store/audio/actions'
import { addTracks } from '../../store/app/slice'
import {
  setSceneAudioPlaylistToggleShuffle,
  setSceneAudioPlaylistChangeRepeat,
  setSceneRemoveAudioPlaylist,
  setSceneAudioPlaylistRemoveAudio,
  setSceneAudioPlaylistSwapAudios
} from '../../store/scene/slice'
import { playAudio } from '../../store/scene/thunks'
import { useAppDispatch, useAppSelector } from '../../store/hooks'
import { orderAudioTags } from '../../store/tag/thunks'
import TagChip from '../library/TagChip'
import {
  selectSceneIsAudioScene,
  selectSceneAudioEnabled,
  selectSceneAudioStartIndex,
  selectSceneAudioPlaylistDuration
} from '../../store/scene/selectors'
import {
  selectAudioTrackNum,
  selectAudioComment,
  selectAudioTags,
  selectAudioThumb,
  selectAudioName,
  selectAudioUrl,
  selectAudioDuration
} from '../../store/audio/selectors'
import { default as AudioPlaylistData } from '../../store/scene/AudioPlaylist'
import flipflip from '../../FlipFlipService'

const useStyles = makeStyles()((theme: Theme) => ({
    audioList: {
      paddingLeft: 0
    },
    mediaIcon: {
      width: '100%',
      height: 'auto'
    },
    thumb: {
      width: theme.spacing(6),
      height: theme.spacing(6)
    },
    playlistAction: {
      textAlign: 'center'
    },
    left: {
      float: 'left',
      paddingLeft: theme.spacing(2)
    },
    right: {
      float: 'right',
      paddingRight: theme.spacing(2)
    },
    trackThumb: {
      height: 40,
      width: 40,
      overflow: 'hidden',
      display: 'flex',
      justifyContent: 'center',
      cursor: 'pointer',
      userSelect: 'none'
    },
    thumbImage: {
      height: '100%'
    },
    listAvatar: {
      width: 56
    },
    bigTooltip: {
      fontSize: 'medium',
      maxWidth: 500
    },
    tagChips: {
      textAlign: 'center'
    },
    avatar: {
      backgroundColor: theme.palette.primary.main,
      boxShadow: 'none'
    },
    sourceIcon: {
      color: theme.palette.primary.contrastText
    }
  }))

interface AudioListItemProps {
  audioID: number
}

function AudioListItem(props: AudioListItemProps) {
  const name = useAppSelector(selectAudioName(props.audioID))
  const thumb = useAppSelector(selectAudioThumb(props.audioID))

  return (
    <ListItem disableGutters>
      <ListItemIcon>
        <Avatar alt={name} src={thumb} className={props.classes.thumb}>
          {thumb == null && (
            <AudiotrackIcon className={props.classes.mediaIcon} />
          )}
        </Avatar>
      </ListItemIcon>
      <ListItemText primary={name} />
    </ListItem>
  )
}

interface PlaylistItemProps {
  audioID: number
  audios: number[]
  removeTrack: (audioID: number) => void
  onSourceOptions: (audioID: number) => void
}

function PlaylistItem(props: PlaylistItemProps) {
  const dispatch = useAppDispatch()
  const trackNum = useAppSelector(selectAudioTrackNum(props.audioID))
  const comment = useAppSelector(selectAudioComment(props.audioID))
  const tags = useAppSelector(selectAudioTags(props.audioID))
  const thumb = useAppSelector(selectAudioThumb(props.audioID))
  const name = useAppSelector(selectAudioName(props.audioID))
  const url = useAppSelector(selectAudioUrl(props.audioID))
  const duration = useAppSelector(selectAudioDuration(props.audioID))

  const onSourceIconClick = (e: MouseEvent<HTMLDivElement>) => {
    const sourceURL = url as string
    if (e.shiftKey && !e.ctrlKey) {
      window.open(url, '_blank')?.focus()
    } else if (!e.shiftKey && e.ctrlKey) {
      flipflip().api.showItemInFolder(sourceURL)
    } else if (!e.shiftKey && !e.ctrlKey) {
      dispatch(playAudio(props.audioID, props.audios))
    }
  }

  const {classes} = useStyles()
  return (
    <ListItem>
      <ListItemAvatar className={classes.listAvatar}>
        <Badge
          invisible={!trackNum}
          max={999}
          overlap="rectangular"
          color="primary"
          badgeContent={trackNum}
        >
          <Tooltip
            disableInteractive
            placement={comment ? 'right' : 'bottom'}
            classes={comment ? { tooltip: classes.bigTooltip } : undefined}
            arrow={!!comment || (tags && tags.length > 0)}
            title={
              comment || (tags && tags.length > 0) ? (
                <div>
                  {comment}
                  {comment && tags && tags.length > 0 && <br />}
                  <div className={classes.tagChips}>
                    {tags &&
                      tags.map((tagID) => (
                        <React.Fragment key={tagID}>
                          <TagChip tagID={tagID} />
                        </React.Fragment>
                      ))}
                  </div>
                </div>
              ) : (
                <div>
                  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Click:
                  Library Tagging
                  <br />
                  Shift+Click: Open Source
                  <br />
                  &nbsp;&nbsp;Ctrl+Click: Reveal File
                </div>
              )
            }
          >
            <div onClick={onSourceIconClick} className={classes.trackThumb}>
              {thumb != null && (
                <img className={classes.thumbImage} src={thumb} />
              )}
              {thumb == null && (
                <Fab size="small" className={classes.avatar}>
                  <SourceIcon url={url} className={classes.sourceIcon} />
                </Fab>
              )}
            </div>
          </Tooltip>
        </Badge>
      </ListItemAvatar>
      <ListItemText primary={name} />
      <ListItemSecondaryAction>
        <Chip
          label={duration && getTimestamp(duration)}
          color="default"
          size="small"
          variant="outlined"
        />
        <IconButton
          edge="end"
          onClick={() => props.onSourceOptions(props.audioID)}
          size="large"
        >
          <BuildIcon />
        </IconButton>
        <IconButton
          edge="end"
          onClick={() => props.removeTrack(props.audioID)}
          size="large"
        >
          <DeleteIcon color={'error'} />
        </IconButton>
      </ListItemSecondaryAction>
    </ListItem>
  )
}

export interface AudioPlaylistProps {
  sceneID: number
  playlistIndex: number
  playlist: AudioPlaylistData
  startPlaying: boolean
  onAddTracks: (playlistIndex: number) => void
  onSourceOptions: (audioID: number) => void
  persist?: boolean
  shorterSeek?: boolean
  showMsTimestamp?: boolean
  scenePaths?: any[]
  goBack?: () => void
  onPlaying?: (position: number, duration: number) => void
  setCurrentAudio?: (audioID: number) => void
}

function AudioPlaylist(props: AudioPlaylistProps) {
  const dispatch = useAppDispatch()
  const isAudioScene = useAppSelector(selectSceneIsAudioScene(props.sceneID))
  const audioEnabled = useAppSelector(selectSceneAudioEnabled(props.sceneID))
  const audioStartIndex = useAppSelector(
    selectSceneAudioStartIndex(props.sceneID)
  )
  const audioPlaylistDuration = useAppSelector(
    selectSceneAudioPlaylistDuration(props.sceneID, props.playlistIndex)
  )

  const [currentIndex, setCurrentIndex] = useState(
    props.playlistIndex === 0 ? audioStartIndex : 0
  )
  const [playingAudios, setPlayingAudios] = useState<number[]>([])

  useEffect(() => {
    if (props.playlistIndex === 0 && isAudioScene) {
      window.addEventListener('keydown', onKeyDown, false)
    }
    restart()

    return () => {
      if (props.playlistIndex === 0 && isAudioScene) {
        window.removeEventListener('keydown', onKeyDown)
      }
    }
  }, [])

  useEffect(() => {
    if (!props.persist) {
      restart()
    }
  }, [props.persist, props.sceneID])

  const restart = () => {
    let audios = props.playlist.audios
    if (props.startPlaying) {
      if (props.playlist.shuffle) {
        audios = randomizeList(Array.from(audios))
      }
      setPlayingAudios(audios)
    }
    if (props.setCurrentAudio) {
      let audio = audios[currentIndex]
      if (!audio) audio = props.playlist.audios[currentIndex]
      props.setCurrentAudio(audio)
    }
  }

  const onKeyDown = (e: KeyboardEvent) => {
    switch (e.key) {
      case '[':
        e.preventDefault()
        dispatch(orderAudioTags(props.playlist.audios[currentIndex]))
        prevTrack()
        break
      case ']':
        e.preventDefault()
        dispatch(orderAudioTags(props.playlist.audios[currentIndex]))
        nextTrack()
        break
    }
  }

  const prevTrack = () => {
    let prevTrack = currentIndex - 1
    if (prevTrack < 0) {
      prevTrack = playingAudios.length - 1
    }
    if (props.setCurrentAudio) {
      props.setCurrentAudio(playingAudios[prevTrack])
    }
    setCurrentIndex(prevTrack)
  }

  const nextTrack = () => {
    let nextTrack = currentIndex + 1
    if (nextTrack >= playingAudios.length) {
      nextTrack = 0
    }
    if (props.setCurrentAudio) {
      props.setCurrentAudio(playingAudios[nextTrack])
    }
    setCurrentIndex(nextTrack)
  }

  const toggleShuffle = () => {
    dispatch(
      setSceneAudioPlaylistToggleShuffle({
        id: props.sceneID,
        value: props.playlistIndex
      })
    )
  }

  const changeRepeat = () => {
    dispatch(
      setSceneAudioPlaylistChangeRepeat({
        id: props.sceneID,
        value: props.playlistIndex
      })
    )
  }

  const removePlaylist = () => {
    dispatch(
      setSceneRemoveAudioPlaylist({
        id: props.sceneID,
        value: props.playlistIndex
      })
    )
  }

  const removeTrack = (audioID: number) => {
    const audioIndex = props.playlist.audios.indexOf(audioID)
    dispatch(
      setSceneAudioPlaylistRemoveAudio({
        id: props.sceneID,
        value: { playlistIndex: props.playlistIndex, audioIndex }
      })
    )
  }

  const {classes} = useStyles()
  if (props.startPlaying) {
    let audioID = playingAudios[currentIndex]
    if (!audioID) audioID = props.playlist.audios[currentIndex]
    if (!audioID) return <div />
    return (
      <React.Fragment>
        <AudioListItem audioID={audioID} classes={classes} />
        <AudioControl
          sceneID={props.sceneID}
          audioID={audioID}
          audioEnabled={audioEnabled || props.persist}
          singleTrack={playingAudios.length === 1}
          lastTrack={currentIndex === playingAudios.length - 1}
          repeat={props.playlist.repeat}
          scenePaths={props.scenePaths}
          shorterSeek={props.shorterSeek}
          showMsTimestamp={props.showMsTimestamp}
          startPlaying={props.startPlaying}
          nextTrack={nextTrack}
          prevTrack={prevTrack}
          onPlaying={props.onPlaying}
          audioVolumeAction={setAudioVolume(playingAudios[currentIndex])}
          goBack={props.goBack}
        />
        <div className={classes.playlistAction}>
          <Tooltip
            disableInteractive
            title={'Shuffle ' + (props.playlist.shuffle ? '(On)' : '(Off)')}
          >
            <IconButton onClick={toggleShuffle} size="large">
              <ShuffleIcon
                color={props.playlist.shuffle ? 'primary' : undefined}
              />
            </IconButton>
          </Tooltip>
          <Tooltip
            disableInteractive
            title={
              'Repeat ' +
              (props.playlist.repeat === RP.none
                ? '(Off)'
                : props.playlist.repeat === RP.all
                  ? '(All)'
                  : '(One)')
            }
          >
            <IconButton onClick={changeRepeat} size="large">
              {props.playlist.repeat === RP.none && <RepeatIcon />}
              {props.playlist.repeat === RP.all && (
                <RepeatIcon color={'primary'} />
              )}
              {props.playlist.repeat === RP.one && (
                <RepeatOneIcon color={'primary'} />
              )}
            </IconButton>
          </Tooltip>
        </div>
      </React.Fragment>
    )
  } else {
    return (
      <List disablePadding>
        <Sortable
          className={classes.audioList}
          options={{
            animation: 150,
            easing: 'cubic-bezier(1, 0, 0, 1)'
          }}
          onChange={(order: any, sortable: any, evt: any) => {
            const newAudios = Array.from(props.playlist.audios)
            arrayMove(newAudios, evt.oldIndex, evt.newIndex)
            dispatch(
              setSceneAudioPlaylistSwapAudios({
                id: props.sceneID,
                value: {
                  playlistIndex: props.playlistIndex,
                  oldIndex: evt.oldIndex,
                  newIndex: evt.newIndex
                }
              })
            )
          }}
        >
          {props.playlist &&
            props.playlist.audios &&
            props.playlist.audios.map((audioID: number, index: number) => (
              <PlaylistItem
                key={index}
                audioID={audioID}
                audios={props.playlist.audios}
                onSourceOptions={props.onSourceOptions}
                removeTrack={removeTrack}
                classes={classes}
              />
            ))}
        </Sortable>
        <div className={classes.playlistAction}>
          <div className={classes.left}>
            <Tooltip
              disableInteractive
              title={'Shuffle ' + (props.playlist.shuffle ? '(On)' : '(Off)')}
            >
              <IconButton onClick={toggleShuffle} size="large">
                <ShuffleIcon
                  color={props.playlist.shuffle ? 'primary' : undefined}
                />
              </IconButton>
            </Tooltip>
            <Tooltip
              disableInteractive
              title={
                'Repeat ' +
                (props.playlist.repeat === RP.none
                  ? '(Off)'
                  : props.playlist.repeat === RP.all
                    ? '(All)'
                    : '(One)')
              }
            >
              <IconButton onClick={changeRepeat} size="large">
                {props.playlist.repeat === RP.none && <RepeatIcon />}
                {props.playlist.repeat === RP.all && (
                  <RepeatIcon color={'primary'} />
                )}
                {props.playlist.repeat === RP.one && (
                  <RepeatOneIcon color={'primary'} />
                )}
              </IconButton>
            </Tooltip>
          </div>
          <Tooltip disableInteractive title="Add Tracks">
            <IconButton
              onClick={() => dispatch(addTracks(props.playlistIndex))}
              size="large"
            >
              <AddIcon />
            </IconButton>
          </Tooltip>
          <div className={classes.right}>
            <Chip
              label={getTimestamp(audioPlaylistDuration)}
              color="default"
              size="small"
              variant="outlined"
            />
            <Tooltip disableInteractive title="Remove Playlist">
              <IconButton onClick={removePlaylist} size="large">
                <ClearIcon color={'error'} />
              </IconButton>
            </Tooltip>
          </div>
        </div>
      </List>
    )
  }
}

;(AudioPlaylist as any).displayName = 'AudioPlaylist'
export default AudioPlaylist
