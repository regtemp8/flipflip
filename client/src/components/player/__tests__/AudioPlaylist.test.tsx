import React from "react";
import { describe, it, expect } from "@jest/globals";
import renderer from "react-test-renderer";
import AudioPlaylist from "../AudioPlaylist";
import TestProvider from "../../../util/TestProvider";
import { RP } from "flipflip-common";
import store from "../../../store/store";
import { newScene } from "../../../store/scene/Scene";
import { setScene } from "../../../store/scene/slice";

jest.mock('../AudioControl', () => 'AudioControl');
jest.mock('../../library/SourceIcon', () => 'SourceIcon');

// mocking this so that test doesn't throw error
jest.mock('react-sortablejs', () => 'Sortable');

describe("AudioPlaylist", () => {
  it("should match snapshot", () => {
    const sceneID = 3
    const playlist = { audios: [], shuffle: false, repeat: RP.none }
    store.dispatch(setScene(newScene({ id: sceneID, audioPlaylists: [playlist] })))
    const component = renderer.create(
      <TestProvider store={store}>
        <AudioPlaylist
          sceneID={sceneID}
          playlistIndex={0}
          playlist={playlist}
          startPlaying={false}
          onAddTracks={(playlistIndex) => {}}
          onSourceOptions={(audio) => {}}
        />
      </TestProvider>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
