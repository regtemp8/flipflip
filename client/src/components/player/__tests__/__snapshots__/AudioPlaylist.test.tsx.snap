// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioPlaylist should match snapshot 1`] = `
<ul
  className="MuiList-root css-1mk9mw3-MuiList-root"
>
  <Sortable
    className="AudioPlaylist-audioList-1"
    onChange={[Function]}
    options={
      Object {
        "animation": 150,
        "easing": "cubic-bezier(1, 0, 0, 1)",
      }
    }
  />
  <div
    className="AudioPlaylist-playlistAction-4"
  >
    <div
      className="AudioPlaylist-left-5"
    >
      <button
        aria-label="Shuffle (Off)"
        aria-labelledby={null}
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element={true}
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ShuffleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10.59 9.17 5.41 4 4 5.41l5.17 5.17 1.42-1.41zM14.5 4l2.04 2.04L4 18.59 5.41 20 17.96 7.46 20 9.5V4h-5.5zm.33 9.41-1.41 1.41 3.13 3.13L14.5 20H20v-5.5l-2.04 2.04-3.13-3.13z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="Repeat (Off)"
        aria-labelledby={null}
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element={true}
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="RepeatIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 7h10v3l4-4-4-4v3H5v6h2V7zm10 10H7v-3l-4 4 4 4v-3h12v-6h-2v4z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <button
      aria-label="Add Tracks"
      aria-labelledby={null}
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
      data-mui-internal-clone-element={true}
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="AddIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      className="AudioPlaylist-right-6"
    >
      <div
        className="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-outlinedDefault css-1biewj9-MuiChip-root"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
      >
        <span
          className="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
        >
          0:00
        </span>
      </div>
      <button
        aria-label="Remove Playlist"
        aria-labelledby={null}
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element={true}
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeMedium css-1wxstni-MuiSvgIcon-root"
          data-testid="ClearIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</ul>
`;
